Marlin/src/feature/easythreed_ui.cpp

```diff
@@ -0,0 +1,242 @@
+        queue.inject(!READ(BTN_RETRACT) ? F("G91\nG0 E10 F180\nG0 E-120 F180\nM104 S0") : F("G91\nG0 E100 F120\nM104 S0"));
+      }
+    } break;
+  }
+
+}
+
+#if HAS_STEPPER_RESET
+  void disableStepperDrivers();
+#endif
+
+//
+// Print Start/Pause/Resume Button
+//
+void EasythreedUI::printButton() {
+  enum KeyStatus : uint8_t { KS_IDLE, KS_PRESS, KS_PROCEED };
+  static uint8_t key_status = KS_IDLE;
+  static millis_t key_time = 0;
+  static bool print_flag = false;
+
+  enum PrintFlag : uint8_t { PF_START, PF_PAUSE, PF_RESUME };
+  static PrintFlag print_key_flag = PF_START;
+
+  const millis_t ms = millis();
+
+  switch (key_status) {
+    case KS_IDLE:
+      if (print_flag && !printingIsActive()) {                      // Reset states if printing state went to inactive
+        print_flag = false;
+        print_key_flag = PF_START;
+        blink_interval_ms = LED_ON;
+      }
+      if (!READ(BTN_PRINT)) {                                       // Print/Pause/Resume button pressed?
+        key_time = ms;                                              // Save start time
+        key_status++;                                               // Go to debounce test
+      }
+      break;
+
+    case KS_PRESS:
+      if (ELAPSED(ms, key_time + BTN_DEBOUNCE_MS))                  // Wait for debounce interval to expire
+        key_status = READ(BTN_PRINT) ? KS_IDLE : KS_PROCEED;        // Proceed if still pressed
+      break;
+
+    case KS_PROCEED:
+      if (!READ(BTN_PRINT)) break;                                  // Wait for the button to be released
+      key_status = KS_IDLE;                                         // Ready for the next press
+      if (PENDING(ms, key_time + 1200 - BTN_DEBOUNCE_MS)) {         // Register a press < 1.2 seconds
+        switch (print_key_flag) {
+          case PF_START:                                            // The "Print" button starts an SD card print
+            if (!printingIsActive()) {                              // Not already printing?
+              card.mount();                                         // Force SD card to mount - now!
+              if (!card.isMounted) {                                // Failed to mount?
+                blink_interval_ms = LED_OFF;                        // Turn off LED
+                print_flag = false;                                 // Clear previous print flag
+                return;                                             // Bail out
+              }
+              card.ls();                                            // List all files to serial output
+              const uint16_t filecnt = card.countFilesInWorkDir();  // Count printable files in cwd
+              if (filecnt == 0) return;                             // None are printable?
+              card.selectFileByIndex(filecnt);                      // Select the last file according to current sort options
```

Comment:
Since file sorting may be alphabetical, reversed, etc., the actual file selected by this method is unpredictable.

Reply:
The intention is to select the most recent file by date/time from the SD card. With the current code this works; but is there a preferred method to achieve this outcome?